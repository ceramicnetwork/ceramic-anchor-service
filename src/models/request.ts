export enum RequestStatus {
  PENDING = 0,
  PROCESSING = 1,
  COMPLETED = 2,
  FAILED = 3,
  READY = 4,
}

export class Request {
  id: number
  status: RequestStatus
  cid: string
  streamId: string
  message: string
  pinned: boolean
  createdAt: Date
  updatedAt: Date
  timestamp: Date
}

export interface RequestUpdateFields {
  message?: string
  status?: RequestStatus
  pinned?: boolean
}

export const REQUEST_MESSAGES = {
  conflictResolutionRejection: 'Request has failed. Updated was rejected by conflict resolution.',
}

export class InvalidRequestStatusError extends Error {
  constructor(status: never) {
    super(`Invalid request status: ${status}`)
  }
}
