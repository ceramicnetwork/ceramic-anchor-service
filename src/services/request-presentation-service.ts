import type { AnchorRepository } from '../repositories/anchor-repository.js'
import type { Request } from '../models/request.js'
import { InvalidRequestStatusError, RequestStatus } from '../models/request.js'
import type { Config } from 'node-config-ts'
import type { IRequestPresentationService } from './request-presentation-service.type.js'

/**
 * Render anchoring Request as JSON for a client to consume.
 */
export class RequestPresentationService implements IRequestPresentationService {
  private readonly schedulerIntervalMS: number

  static inject = ['config', 'anchorRepository'] as const

  constructor(config: Config, private readonly anchorRepository: AnchorRepository) {
    this.schedulerIntervalMS = config.schedulerIntervalMS
  }

  /**
   * Rich JSON of a request.
   *
   * @param request - Request to be rendered as JSON.
   */
  async body(request: Request): Promise<any> {
    switch (request.status) {
      case RequestStatus.COMPLETED: {
        const anchor = await this.anchorRepository.findByRequest(request)
        // TODO: This is a workaround, fix in CDB-2192
        const anchorCommit = {
          cid: anchor ? anchor.cid : request.cid,
          content: {
            // okay to be undefined because it is not used by ceramic node
            path: anchor?.path,
            prev: request.cid,
            // okay to be undefined because it is not used by ceramic node
            proof: anchor?.proofCid,
          },
        }

        return {
          id: request.id,
          status: RequestStatus[request.status],
          cid: request.cid,
          docId: request.streamId, // todo remove
          streamId: request.streamId,
          message: request.message,
          createdAt: request.createdAt.getTime(),
          updatedAt: request.updatedAt.getTime(),
          // TODO: Remove this backwards compatibility field
          anchorRecord: anchorCommit,
          anchorCommit,
        }
      }
      case RequestStatus.PENDING: {
        return {
          id: request.id,
          status: RequestStatus[request.status],
          cid: request.cid,
          docId: request.streamId, // TODO remove
          streamId: request.streamId,
          message: request.message,
          createdAt: request.createdAt.getTime(),
          updatedAt: request.updatedAt.getTime(),
        }
      }
      case RequestStatus.PROCESSING:
        return {
          id: request.id,
          status: RequestStatus[request.status],
          cid: request.cid,
          docId: request.streamId, // TODO remove
          streamId: request.streamId,
          message: request.message,
          createdAt: request.createdAt.getTime(),
          updatedAt: request.updatedAt.getTime(),
        }
      case RequestStatus.FAILED:
        return {
          id: request.id,
          status: RequestStatus[request.status],
          cid: request.cid,
          docId: request.streamId, // TODO remove
          streamId: request.streamId,
          message: request.message,
          createdAt: request.createdAt.getTime(),
          updatedAt: request.updatedAt.getTime(),
        }
      case RequestStatus.READY:
        return {
          id: request.id,
          status: RequestStatus[request.status],
          cid: request.cid,
          docId: request.streamId, // TODO remove
          streamId: request.streamId,
          message: request.message,
          createdAt: request.createdAt.getTime(),
          updatedAt: request.updatedAt.getTime(),
        }
      default:
        throw new InvalidRequestStatusError(request.status)
    }
  }
}
