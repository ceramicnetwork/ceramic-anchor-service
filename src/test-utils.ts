import CID from 'cids'
import { CeramicService } from './services/ceramic-service'
import { IpfsService } from './services/ipfs-service'
import { StreamID, CommitID } from '@ceramicnetwork/streamid'
import { AnchorCommit, MultiQuery, Stream } from '@ceramicnetwork/common'
import dagCBOR from 'ipld-dag-cbor'
import { randomBytes } from '@stablelib/random'

export async function randomCID(): Promise<CID> {
  return dagCBOR.util.cid(randomBytes(32))
}

export class MockIpfsService implements IpfsService {
  private _streams: Record<string, any> = {}

  constructor() {}

  async init(): Promise<void> {
    return null
  }

  async retrieveRecord(cid: CID | string): Promise<any> {
    return this._streams[cid.toString()]
  }

  async storeRecord(record: Record<string, unknown>): Promise<CID> {
    const cid = await randomCID()
    this._streams[cid.toString()] = record
    return cid
  }

  reset() {
    this._streams = {}
  }
}

export class MockCeramicService implements CeramicService {
  constructor(
    private _ipfsService: IpfsService,
    private _streams: Record<string, any> = {},
    private _cidIndex = 0
  ) {}

  async loadStream(streamId: StreamID): Promise<any> {
    return this._streams[streamId.toString()]
  }

  async multiQuery(queries: MultiQuery[]): Promise<Record<string, Stream>> {
    const result = {}
    for (const query of queries) {
      const id = query.streamId.toString()
      const stream = this._streams[id]
      if (stream) {
        result[id] = stream
      }
    }

    return result
  }

  async publishAnchorCommit(streamId: StreamID, anchorCommit: AnchorCommit): Promise<CID> {
    return this._ipfsService.storeRecord(anchorCommit)
  }

  // Mock-only method to control what gets returned by loadStream()
  putStream(id: StreamID | CommitID, stream: any) {
    this._streams[id.toString()] = stream
  }

  // Mock-only method to generate a random base StreamID
  async generateBaseStreamID(): Promise<StreamID> {
    const cid = await randomCID()
    return new StreamID('tile', cid)
  }

  reset() {
    this._cidIndex = 0
    this._streams = {}
  }
}
