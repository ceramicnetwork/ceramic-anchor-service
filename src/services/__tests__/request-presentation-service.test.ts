import { describe, expect, jest, test } from '@jest/globals'
import { RequestStatus } from '../../models/request.js'
import { RequestPresentationService } from '../request-presentation-service.js'
import type {
  AnchorWithRequest,
  IAnchorRepository,
} from '../../repositories/anchor-repository.type.js'
import { generateRequest } from '../../__tests__/test-utils.js'

const anchorRepository = {
  findByRequest: jest.fn(),
} as unknown as IAnchorRepository

const service = new RequestPresentationService(anchorRepository)

const REQUEST_OVERRIDE = {
  id: '889483296',
  cid: 'bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m',
  streamId: 'k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72',
  message: 'Fresh request',
  createdAt: new Date('2020-01-02T03:04Z'),
  updatedAt: new Date('2021-02-03T04:05Z'),
}

describe('present by RequestStatus', () => {
  test('PENDING, PROCESSING, FAILED, READY', async () => {
    const statuses = [
      RequestStatus.PENDING,
      RequestStatus.PROCESSING,
      RequestStatus.FAILED,
      RequestStatus.READY,
    ]
    for (const status of statuses) {
      const request = generateRequest({
        ...REQUEST_OVERRIDE,
        status: status,
      })
      const presentation = await service.body(request)
      expect(presentation).toMatchSnapshot()
    }
  })

  test('REPLACED', async () => {
    const request = generateRequest({
      ...REQUEST_OVERRIDE,
      status: RequestStatus.REPLACED,
    })
    const presentation = await service.body(request)
    // REPLACED is an internal status, which translates to "FAILED" externally
    expect(presentation.status).toEqual(RequestStatus[RequestStatus.FAILED])
    expect(presentation).toMatchSnapshot()
  })

  test('COMPLETED', async () => {
    const request = generateRequest({
      ...REQUEST_OVERRIDE,
      status: RequestStatus.COMPLETED,
    })
    const findByRequestSpy = jest.spyOn(anchorRepository, 'findByRequest')
    const anchor = {
      path: '/some/path',
      cid: 'anchor-cid',
      proofCid: 'anchor-proof-cid',
    }
    findByRequestSpy.mockImplementationOnce(async () => {
      return { ...anchor, request: request } as AnchorWithRequest
    })
    const presentation = await service.body(request)
    expect(presentation).toMatchSnapshot()
  })

  test('COMPLETED but no associated anchor', async () => {
    const request = generateRequest({
      ...REQUEST_OVERRIDE,
      status: RequestStatus.COMPLETED,
    })
    const findByRequestSpy = jest.spyOn(anchorRepository, 'findByRequest')
    findByRequestSpy.mockImplementationOnce(async () => {
      return null
    })
    const presentation = await service.body(request)
    expect(presentation).toMatchSnapshot()
  })
})
