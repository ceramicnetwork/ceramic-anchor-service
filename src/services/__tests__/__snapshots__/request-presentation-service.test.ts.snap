// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`present by RequestStatus COMPLETED 1`] = `
{
  "anchorCommit": {
    "cid": "anchor-cid",
    "content": {
      "path": "/some/path",
      "prev": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
      "proof": "anchor-proof-cid",
    },
  },
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "COMPLETED",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus COMPLETED but no associated anchor 1`] = `
{
  "anchorCommit": {
    "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
    "content": {
      "path": undefined,
      "prev": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
      "proof": undefined,
    },
  },
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "COMPLETED",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus PENDING, PROCESSING, FAILED, READY 1`] = `
{
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "PENDING",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus PENDING, PROCESSING, FAILED, READY 2`] = `
{
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "PROCESSING",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus PENDING, PROCESSING, FAILED, READY 3`] = `
{
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "FAILED",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus PENDING, PROCESSING, FAILED, READY 4`] = `
{
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "READY",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;

exports[`present by RequestStatus REPLACED 1`] = `
{
  "cid": "bafyreibfyl5p56xjdarie2p7brjmwktxgiggdm6hxyeugauk6zxg5c6g6m",
  "createdAt": 1577934240000,
  "id": "889483296",
  "message": "Fresh request",
  "status": "FAILED",
  "streamId": "k2t6wyfsu4pfxu08vo93w38oyu9itsuf374ekyeno0wb62ozm2o9sznrn8qp72",
  "updatedAt": 1612325100000,
}
`;
